streams/src/main/java/org/apache/kafka/streams/processor/internals/DefaultStateUpdater.java

```diff
@@ -500,23 +500,29 @@ public void remove(final TaskId taskId) {
         }
     }
 
-    @Override
-    public void pause(final TaskId taskId) {
+    public void pause(final TopologyMetadata topologyMetadata) {
         tasksAndActionsLock.lock();
         try {
-            tasksAndActions.add(TaskAndAction.createPauseTask(taskId));
-            tasksAndActionsCondition.signalAll();
+            for (final Task task : getUpdatingTasks()) {
+                if (topologyMetadata.isPaused(task.id().topologyName())) {
+                    tasksAndActions.add(TaskAndAction.createPauseTask(task.id()));
+                    tasksAndActionsCondition.signalAll();
+                }
+            }
```

Comment:
Wouldn't it be better to store paused topologies within the state updater thread? 

Consider the following scenario: 
1. The state updater managed task A from topology X. 
2. Topology X is paused.
3. Task A is revoked from and task B from topology X is assigned.
When the topology is paused, we would pass the topology name through an input queue event (like add or remove) to the state updater thread. Once the state updater thread processes the input queue event with the topology name to pause, it stores the name of the paused topology and pauses all tasks (i.e., task A) of the paused topology. Once task B is assigned and added to the state updater, the paused topologies are consulted and since topology X is paused task B will be paused directly. When a topology is resumed, the corresponding tasks are resumed  and the name of the resumed topology is removed from the state updater thread.
If we expose the names of the paused topology from the state updater thread to the default state updater, we can check in `pause(topology)` if the state updater thread already knows about the paused topology and avoid creating an event in the input queue for already-known paused topologies. 
All this would avoid to iterate over the tasks each time `checkStateUpdater` is called. We would only do a check of the paused topologies when a task is added, and loop over the task if a topology was just paused.

WDYT?

Reply:
Hi @cadonna I tried out this idea but it turns out more complicated than expected. The main issue is that if we store the paused topology outside the restore thread, and only let the caller thread to modify it, the restore thread still need to access it while handling add-task; and vice versa if we store the paused topologies inside the restore thread and let that thread to modify it upon handling pause/resume-task, the caller thread still need to access it to determine if a new action needs to be enqueued. In either case we'd construct two communication channels between the two.

In the end I chose a different route, which is directly pass in the `topologyMetadata` to StateUpdater, and then periodically check that topology's paused metadata to pause / resume tasks. As a result we would also remove pause/resume-task actions as a queue item. Right now I just piggy-backed it with checkpointing but I'm okay with doing it separately, periodically. After some pondering I feel this is actually simpler than passing pause/resume actions via channels.

Reply:
@guozhangwang I think, we misunderstood each other. Sorry if I was not clear enough.
My proposal was to store the names of the paused topologies in the state updater thread, like:

```
public class DefaultStateUpdater implements StateUpdater {
...
    private class StateUpdaterThread extends Thread {
    ...
        private Set<String> pausedTopologies = ConcurrentHashMap.newKeySet();

        private Set<String> getPausedTopologies() {
            return Collections.unmodifiableSet(pausedTopologies);
        }
```

The `StateUpdater` (and `DefaultStateUpdater`) would then have the following method:

```
public class DefaultStateUpdater implements StateUpdater {
...
    @Override
    public syncPausedTopologies(final Set<String> pausedTopologies) {
        final Set<String> stateUpdaterPausedTopologies = getPausedTopologies();
        final Set<String> pausedTopologiesToAdd = pausedTopologies.removeAll(stateUpdaterPausedTopologies);
        final Set<String> pausedTopologiesToRemove = stateUpdaterPausedTopologies.removeAll(pausedTopologies);
        if (!pausedTopologiesToAdd.isEmpty()) {  
            tasksAndActions.addPausedTopology(TaskAndAction.createAddPausedTopology(pausedTopologiesToAdd));
        }
        if (!pausedTopologiesToRemove.isEmpty()) {  
            tasksAndActions.removePausedTopology(TaskAndAction.createRemovePausedTopology(pausedTopologiesToRemove));
        }
    }
``` 

In this way, we would not need to iterate over the tasks in the state updater in regular intervals but only if the set of paused topologies changed. 
When we add a task to the state updater, the state updater thread needs to check if the topology of the task is paused or not.

```
public class DefaultStateUpdater implements StateUpdater {
...
    private class StateUpdaterThread extends Thread {
    ...
        private void addTask(final Task task) {
             if (isStateless(task)) {
                addToRestoredTasks((StreamTask) task);
                log.info("Stateless active task " + task.id() + " was added to the restored tasks of the state updater");
            } else {
                if (pausedTopologies.contains(task.topologyName())) {
                    // add to paused tasks
                } else {
                   // add to updating tasks 
                }
            }
```

Reply:
Hi @cadonna I actually did exactly this with a concurrent hash map keysets :) The main issue is because the `pausedTopologies` would be modified by the main thread, while it would be checked upon `add` by the restore thread, there's a race condition complexity here since the main thread has a sequence of 1) check the paused topology concurrent set, 2) create pause/resume tasks into the queue, 3) update the paused topology concurrent set. And with restore thread checking the paused topology set concurrently, we could have many racing complexities unless we synchronize the whole section of 1/2/3 steps above. It's definitely doable, but it's introducing a lot of edge cases handling when getting an action from the queue.

Introducing the whole `topology metadata` into the state updater on the other hand could delay the syncing of paused topologies as we only check it infrequently, but we are asynchronously updating the paused topologies anyways, and I feel its less complex.